src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -149,45 +193,116 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     SetSslCertificate(innerContext, certHandle!, certKeyHandle!);
                 }
 
-                if (sslAuthenticationOptions.IsServer && sslAuthenticationOptions.RemoteCertRequired)
+            }
+            catch
+            {
+                innerContext.Dispose();
+                throw;
+            }
+
+            return innerContext;
+        }
+
+        // This essentially wraps SSL* SSL_new()
+        internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credential, SslAuthenticationOptions sslAuthenticationOptions)
+        {
+            SafeSslHandle? context = null;
+            SafeSslContextHandle? sslCtx = null;
+            SafeSslContextHandle? innerContext = null;
```

Comment:
OK, having gotten to the end I think that `innerContext` is "an SSL_CTX that got newly allocated" (not from the cache) and `sslCtx` is "whether cached or newly allocated, it's the one we're using".

It looks like `innerContext` could be replaced with `bool createdNewCtx = false;`, and then things would probably be more clear.